### PR TITLE
Optimize NullableSingleInputAggregationFunction when entire block is null

### PR DESCRIPTION
**Pull request by yashmayya**
- Currently, if null handling is enabled and any aggregation function extending `NullableSingleInputAggregationFunction` encounters a block where every value is `null`, it will unnecessarily iterate through every entry in the null bitmap with a check to see whether any non-null values were encountered in each iteration.
- Instead, we can skip over this completely based on the null bitmap's cardinality. This optimization already exists in some of the other aggregation functions that don't currently extend `NullableSingleInputAggregationFunction` - 

https://github.com/apache/pinot/blob/0f2dcb770dca414b02dc4d495d4b4458fa2e91ec/pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/SumAggregationFunction.java#L135

https://github.com/apache/pinot/blob/0f2dcb770dca414b02dc4d495d4b4458fa2e91ec/pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AvgAggregationFunction.java#L102
- Some of the existing unit tests such as `ModeAggregationFunctionTest` and `AbstractPercentileAggregationFunctionTest` already provide coverage for this scenario.

### CODE REVIEW

**Review by Jackie-Jiang — COMMENTED** [review]
Not introduced in this PR: do we have a contract that the null bitmap can never be empty? i.e. either `null` or non-empty?

**Comment by richardstartin** [comment]
To extend this idea, it would be better to check if the bitmap is empty with the isEmpty method (constant time) rather than compute its cardinality (worst case linear time).

**Comment by Jackie-Jiang** [comment]
@richardstartin Good point! Is there an API to check if the bitmap is all set? Ideally we want to early terminate whenever there is an unset bit

**Comment by yashmayya** [comment]
> do we have a contract that the null bitmap can never be empty

It does look like the `NullValueVectorCreator` only creates the null value vector file on disk if the bitmap is not empty - https://github.com/apache/pinot/blob/ee4049e1e80bc2521242dd523236f81549e03222/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/nullvalue/NullValueVectorCreator.java#L69-L78

And even though the mutable null value vector reader could return an empty bitmap, I see some places like in `ProjectionBlockValSet` where that is handled and a null bitmap is returned if empty - https://github.com/apache/pinot/blob/ee4049e1e80bc2521242dd523236f81549e03222/pinot-core/src/main/java/org/apache/pinot/core/operator/docvalsets/ProjectionBlockValSet.java#L65-L67

Seems like an implicit contract but I guess it shouldn't be much of an issue either way since the `RoaringBitmap::isEmpty` method is O(1) time complexity?

> rather than compute its cardinality (worst case linear time)

Ah thanks, I hadn't looked too closely at the implementation and didn't realize that it could be worst case linear time. In that case, I'm not sure if this optimization makes much sense.

**Comment by gortiz** [comment]
> Ah thanks, I hadn't looked too closely at the implementation and didn't realize that it could be worst case linear time. In that case, I'm not sure if this optimization makes much sense.

Can you use my branch with the benchmarks to compare performance?

**Comment by yashmayya** [comment]
@gortiz I used a modified version of the mode aggregation benchmark (with some changes to add a case where all values in the block are `null` and deterministic null handling enabled / disabled) that you added in https://github.com/apache/pinot/pull/12227/ and here are the results:

Without the optimization from this PR (`_nullPeriod` = `1` implies all the values in a block are `null`):

```
Benchmark                      (_nullHandlingEnabled)  (_nullPeriod)   Mode  Cnt   Score   Error   Units
BenchmarkModeAggregation.test                    true              1  thrpt   50  46.623 ± 0.180  ops/ms
BenchmarkModeAggregation.test                    true              2  thrpt   50  36.023 ± 0.185  ops/ms
BenchmarkModeAggregation.test                    true              4  thrpt   50  32.064 ± 0.142  ops/ms
BenchmarkModeAggregation.test                    true              8  thrpt   50  31.526 ± 0.212  ops/ms
BenchmarkModeAggregation.test                    true             16  thrpt   50  31.196 ± 0.443  ops/ms
BenchmarkModeAggregation.test                    true             32  thrpt   50  32.062 ± 0.198  ops/ms
BenchmarkModeAggregation.test                    true             64  thrpt   50  31.992 ± 0.215  ops/ms
BenchmarkModeAggregation.test                    true            128  thrpt   50  32.423 ± 0.166  ops/ms
BenchmarkModeAggregation.test                   false              1  thrpt   50  28.642 ± 0.380  ops/ms
BenchmarkModeAggregation.test                   false              2  thrpt   50  27.387 ± 0.258  ops/ms
BenchmarkModeAggregation.test                   false              4  thrpt   50  28.230 ± 0.406  ops/ms
BenchmarkModeAggregation.test                   false              8  thrpt   50  27.588 ± 0.335  ops/ms
BenchmarkModeAggregation.test                   false             16  thrpt   50  27.950 ± 0.275  ops/ms
BenchmarkModeAggregation.test                   false             32  thrpt   50  27.816 ± 0.480  ops/ms
BenchmarkModeAggregation.test                   false             64  thrpt   50  28.263 ± 0.247  ops/ms
BenchmarkModeAggregation.test                   false            128  thrpt   50  27.943 ± 0.175  ops/ms
```

With the optimization from this PR:

```
Benchmark                      (_nullHandlingEnabled)  (_nullPeriod)   Mode  Cnt   Score   Error   Units
BenchmarkModeAggregation.test                    true              1  thrpt   50  154024.598 ± 641.790  ops/ms
BenchmarkModeAggregation.test                    true              2  thrpt   50  36.364 ± 0.209  ops/ms
BenchmarkModeAggregation.test                    true              4  thrpt   50  32.198 ± 0.148  ops/ms
BenchmarkModeAggregation.test                    true              8  thrpt   50  29.141 ± 0.235  ops/ms
BenchmarkModeAggregation.test                    true             16  thrpt   50  31.579 ± 0.272  ops/ms
BenchmarkModeAggregation.test                    true             32  thrpt   50  32.216 ± 0.154  ops/ms
BenchmarkModeAggregation.test                    true             64  thrpt   50  32.227 ± 0.128  ops/ms
BenchmarkModeAggregation.test                    true            128  thrpt   50  32.264 ± 0.130  ops/ms
BenchmarkModeAggregation.test                   false              1  thrpt   50  27.662 ± 0.308  ops/ms
BenchmarkModeAggregation.test                   false              2  thrpt   50  28.190 ± 0.161  ops/ms
BenchmarkModeAggregation.test                   false              4  thrpt   50  27.893 ± 0.180  ops/ms
BenchmarkModeAggregation.test                   false              8  thrpt   50  28.095 ± 0.144  ops/ms
BenchmarkModeAggregation.test                   false             16  thrpt   50  28.844 ± 0.619  ops/ms
BenchmarkModeAggregation.test                   false             32  thrpt   50  28.177 ± 0.293  ops/ms
BenchmarkModeAggregation.test                   false             64  thrpt   50  28.170 ± 0.169  ops/ms
BenchmarkModeAggregation.test                   false            128  thrpt   50  28.215 ± 0.298  ops/ms
```

So the cases where not all values in a block are null are basically unaffected but there is indeed a huge increase (~3000X) in the throughput for the all nulls case.

**Comment by yashmayya** [comment]
I've updated the mode aggregation benchmark and pushed the change to this branch itself. Here are the results:

Without this optimization:
```
Benchmark                      (_nullHandlingEnabledPerCent)  (_nullPeriod)   Mode  Cnt   Score   Error   Units
BenchmarkModeAggregation.test                            100              1  thrpt   50  46.303 ± 0.274  ops/ms
BenchmarkModeAggregation.test                            100              2  thrpt   50  36.063 ± 0.203  ops/ms
BenchmarkModeAggregation.test                            100              4  thrpt   50  31.343 ± 0.128  ops/ms
BenchmarkModeAggregation.test                            100              8  thrpt   50  30.965 ± 0.631  ops/ms
BenchmarkModeAggregation.test                            100             16  thrpt   50  31.089 ± 0.154  ops/ms
BenchmarkModeAggregation.test                            100             32  thrpt   50  32.219 ± 0.192  ops/ms
BenchmarkModeAggregation.test                            100             64  thrpt   50  32.074 ± 0.157  ops/ms
BenchmarkModeAggregation.test                            100            128  thrpt   50  32.154 ± 0.388  ops/ms
BenchmarkModeAggregation.test                             50              1  thrpt   50  47.861 ± 8.267  ops/ms
BenchmarkModeAggregation.test                             50              2  thrpt   50  34.433 ± 0.933  ops/ms
BenchmarkModeAggregation.test                             50              4  thrpt   50  31.789 ± 0.204  ops/ms
BenchmarkModeAggregation.test                             50              8  thrpt   50  32.006 ± 0.303  ops/ms
BenchmarkModeAggregation.test                             50             16  thrpt   50  32.274 ± 0.243  ops/ms
BenchmarkModeAggregation.test                             50             32  thrpt   50  32.425 ± 0.200  ops/ms
BenchmarkModeAggregation.test                             50             64  thrpt   50  32.301 ± 0.171  ops/ms
BenchmarkModeAggregation.test                             50            128  thrpt   50  32.473 ± 0.147  ops/ms
BenchmarkModeAggregation.test                              0              1  thrpt   50  28.122 ± 0.310  ops/ms
BenchmarkModeAggregation.test                              0              2  thrpt   50  27.857 ± 0.163  ops/ms
BenchmarkModeAggregation.test                              0              4  thrpt   50  27.667 ± 0.164  ops/ms
BenchmarkModeAggregation.test                              0              8  thrpt   50  27.933 ± 0.188  ops/ms
BenchmarkModeAggregation.test                              0             16  thrpt   50  27.772 ± 0.159  ops/ms
BenchmarkModeAggregation.test                              0             32  thrpt   50  28.018 ± 0.187  ops/ms
BenchmarkModeAggregation.test                              0             64  thrpt   50  28.060 ± 0.358  ops/ms
BenchmarkModeAggregation.test                              0            128  thrpt   50  28.159 ± 0.192  ops/ms
```

With this optimization:
```
Benchmark                      (_nullHandlingEnabledPerCent)  (_nullPeriod)   Mode  Cnt       Score       Error   Units
BenchmarkModeAggregation.test                            100              1  thrpt   50  154371.180 ±   592.954  ops/ms
BenchmarkModeAggregation.test                            100              2  thrpt   50      36.142 ±     0.301  ops/ms
BenchmarkModeAggregation.test                            100              4  thrpt   50      32.071 ±     0.403  ops/ms
BenchmarkModeAggregation.test                            100              8  thrpt   50      31.035 ±     0.396  ops/ms
BenchmarkModeAggregation.test                            100             16  thrpt   50      31.827 ±     0.201  ops/ms
BenchmarkModeAggregation.test                            100             32  thrpt   50      32.297 ±     0.138  ops/ms
BenchmarkModeAggregation.test                            100             64  thrpt   50      32.251 ±     0.158  ops/ms
BenchmarkModeAggregation.test                            100            128  thrpt   50      32.452 ±     0.161  ops/ms
BenchmarkModeAggregation.test                             50              1  thrpt   50   67106.709 ± 37843.797  ops/ms
BenchmarkModeAggregation.test                             50              2  thrpt   50      34.638 ±     0.855  ops/ms
BenchmarkModeAggregation.test                             50              4  thrpt   50      32.408 ±     0.210  ops/ms
BenchmarkModeAggregation.test                             50              8  thrpt   50      29.867 ±     0.635  ops/ms
BenchmarkModeAggregation.test                             50             16  thrpt   50      31.973 ±     0.408  ops/ms
BenchmarkModeAggregation.test                             50             32  thrpt   50      32.236 ±     0.293  ops/ms
BenchmarkModeAggregation.test                             50             64  thrpt   50      31.478 ±     0.494  ops/ms
BenchmarkModeAggregation.test                             50            128  thrpt   50      32.253 ±     0.276  ops/ms
BenchmarkModeAggregation.test                              0              1  thrpt   50      27.938 ±     0.179  ops/ms
BenchmarkModeAggregation.test                              0              2  thrpt   50      27.863 ±     0.204  ops/ms
BenchmarkModeAggregation.test                              0              4  thrpt   50      28.036 ±     0.202  ops/ms
BenchmarkModeAggregation.test                              0              8  thrpt   50      28.203 ±     0.372  ops/ms
BenchmarkModeAggregation.test                              0             16  thrpt   50      28.106 ±     0.146  ops/ms
BenchmarkModeAggregation.test                              0             32  thrpt   50      28.663 ±     0.352  ops/ms
BenchmarkModeAggregation.test                              0             64  thrpt   50      28.048 ±     0.145  ops/ms
BenchmarkModeAggregation.test                              0            128  thrpt   50      28.103 ±     0.164  ops/ms
```

Also, I tried this optimization with a couple of other aggregation functions and benchmarks locally and saw similar result.

**Review by gortiz — APPROVED** [review]
It looks good to me, but lets wait for the review from @vvivekiyer given @siddharthteotia asked for that

**Comment by yashmayya** [comment]
@vvivekiyer @siddharthteotia can this be merged? It's a small optimization but with a huge performance boost in a very specific scenario.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:84]: As @richardstartin said, `getCardinality()` can be linear time complexity in the worst case. Here we want to check if the bitmap is fully set, and a cheaper method for the same purpose is `roaringBitmap.contains(0, length)` which early terminate when it detects an unset bit.

We are doing similar checks in quite some places. Do you want to merge this one first, then open a separate PR to fix all of them?

Comment (yashmayya) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:84]: Thanks, that's a good call out, I hadn't looked at the `contains` API for this use case and I can confirm that it does indeed perform slightly better than `getCardinality` from this PR in the `_nullPeriod` = 1 case (and with similar performance in other cases). Results with `contains`:

```
Benchmark                      (_nullHandlingEnabledPerCent)  (_nullPeriod)   Mode  Cnt       Score       Error   Units
BenchmarkModeAggregation.test                            100              1  thrpt   50  174354.600 ±  1313.033  ops/ms
BenchmarkModeAggregation.test                            100              2  thrpt   50      36.165 ±     0.248  ops/ms
BenchmarkModeAggregation.test                            100              4  thrpt   50      32.483 ±     0.106  ops/ms
BenchmarkModeAggregation.test                            100              8  thrpt   50      31.178 ±     0.124  ops/ms
BenchmarkModeAggregation.test                            100             16  thrpt   50      31.567 ±     0.446  ops/ms
BenchmarkModeAggregation.test                            100             32  thrpt   50      31.649 ±     0.447  ops/ms
BenchmarkModeAggregation.test                            100             64  thrpt   50      32.200 ±     0.134  ops/ms
BenchmarkModeAggregation.test                            100            128  thrpt   50      32.311 ±     0.177  ops/ms
BenchmarkModeAggregation.test                             50              1  thrpt   50   73644.879 ± 41534.000  ops/ms
BenchmarkModeAggregation.test                             50              2  thrpt   50      33.695 ±     1.105  ops/ms
BenchmarkModeAggregation.test                             50              4  thrpt   50      32.220 ±     0.207  ops/ms
BenchmarkModeAggregation.test                             50              8  thrpt   50      31.126 ±     0.447  ops/ms
BenchmarkModeAggregation.test                             50             16  thrpt   50      31.884 ±     0.296  ops/ms
BenchmarkModeAggregation.test                             50             32  thrpt   50      31.326 ±     0.591  ops/ms
BenchmarkModeAggregation.test                             50             64  thrpt   50      32.018 ±     0.286  ops/ms
BenchmarkModeAggregation.test                             50            128  thrpt   50      32.320 ±     0.190  ops/ms
BenchmarkModeAggregation.test                              0              1  thrpt   50      27.950 ±     0.199  ops/ms
BenchmarkModeAggregation.test                              0              2  thrpt   50      27.595 ±     0.145  ops/ms
BenchmarkModeAggregation.test                              0              4  thrpt   50      27.524 ±     0.160  ops/ms
BenchmarkModeAggregation.test                              0              8  thrpt   50      28.522 ±     0.266  ops/ms
BenchmarkModeAggregation.test                              0             16  thrpt   50      27.813 ±     0.167  ops/ms
BenchmarkModeAggregation.test                              0             32  thrpt   50      28.266 ±     0.424  ops/ms
BenchmarkModeAggregation.test                              0             64  thrpt   50      27.646 ±     0.295  ops/ms
BenchmarkModeAggregation.test                              0            128  thrpt   50      27.856 ±     0.173  ops/ms
```

(results from `getCardinality` are in [this comment](https://github.com/apache/pinot/pull/13758#issuecomment-2276020558))

> Do you want to merge this one first, then open a separate PR to fix all of them?

Yes, I'd be happy to file a follow-up PR to update it in all the places 👍

Comment (yashmayya) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java]: Actually, on second thought, the other places should be updated to use `NullableSingleInputAggregationFunction` in https://github.com/apache/pinot/pull/13791 anyway. So I've pushed the change to use `roaringBitmap.contains(0, length)` instead of `roaringBitmap.getCardinality() >= length` to  this PR itself.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java:129]: (minor) We might want to move this before getting the iterator

**Review by Jackie-Jiang — APPROVED** [review]
LGTM

**Pull request merged**
